homeassistant/components/sma/config_flow.py

```diff
@@ -0,0 +1,215 @@
+
+    if await sma.new_session() is False:
+        raise InvalidAuth
+
+    device_info = await sma.device_info()
+
+    if not device_info:
+        raise CannotRetrieveDeviceInfo
+
+    await sma.close_session()
+    return device_info
+
+
+class SmaConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for SMA."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize."""
+        self._data = {
+            CONF_HOST: vol.UNDEFINED,
+            CONF_SSL: False,
+            CONF_VERIFY_SSL: True,
+            CONF_GROUP: GROUPS[0],
+            CONF_PASSWORD: vol.UNDEFINED,
+            CONF_SENSORS: [],
+            CONF_CUSTOM: {},
+            DEVICE_INFO: {},
+        }
+
+    async def async_step_user(self, user_input=None):
+        """First step in config flow."""
+        errors = {}
+        if user_input is not None:
+            self._data[CONF_HOST] = user_input[CONF_HOST]
+            self._data[CONF_SSL] = user_input[CONF_SSL]
+            self._data[CONF_VERIFY_SSL] = user_input[CONF_VERIFY_SSL]
+            self._data[CONF_GROUP] = user_input[CONF_GROUP]
+            self._data[CONF_PASSWORD] = user_input[CONF_PASSWORD]
+
+            try:
+                self._data[DEVICE_INFO] = await validate_input(self.hass, user_input)
+            except aiohttp.ClientError:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except CannotRetrieveDeviceInfo:
+                errors["base"] = "cannot_retrieve_device_info"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            if not errors:
+                await self.async_set_unique_id(self._data[DEVICE_INFO]["serial"])
+                self._abort_if_unique_id_configured()
+                return await self.async_step_sensors()
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST, default=self._data[CONF_HOST]): cv.string,
+                    vol.Optional(CONF_SSL, default=self._data[CONF_SSL]): cv.boolean,
+                    vol.Optional(
+                        CONF_VERIFY_SSL, default=self._data[CONF_VERIFY_SSL]
+                    ): cv.boolean,
+                    vol.Optional(CONF_GROUP, default=self._data[CONF_GROUP]): vol.In(
+                        GROUPS
+                    ),
+                    vol.Required(
+                        CONF_PASSWORD, default=self._data[CONF_PASSWORD]
+                    ): cv.string,
+                }
+            ),
+            errors=errors,
+        )
+
+    async def async_step_sensors(self, user_input=None):
+        """Second step in config flow to select sensors to create."""
```

Comment:
When adding a new device, we only want to collect the absolutely minimum amount of information to get the device setup to reduce the burden on the user to understand what the need to do to integrate their device.

Since they have unique ids and can be disabled in the UI, can we create them all and let the user disable the ones they don't want via the standard methods?

Reply:
I modeled the config flow after the yaml config. In yaml you were able to select the sensors to add.
The library fetches only the data of the selected sensors. I could remove this step and add all sensors by default.

I assume disable entities will not be updated and data does not have to be fetched? Probably have to make a custom data coordinator to keep track of, and only fetch information for, the enabled sensors. Do you know of any good examples?

Reply:
`august` has a `subscriber.py` that can be used a a base class to accomplish this.

Reply:
I have found a cleaner solution by enabling or disabling the sensor in the pysma library.  See https://github.com/kellerza/pysma/pull/55 and 

Now the entities can be disable/enabled I removed the `async_step_sensors()` in 4081d84